source/extensions/load_balancing_policies/subset/subset_lb_config.cc

```diff
@@ -0,0 +1,106 @@
+#include "source/extensions/load_balancing_policies/subset/subset_lb_config.h"
+
+#include "source/common/config/utility.h"
+#include "source/common/upstream/upstream_impl.h"
+
+namespace Envoy {
+namespace Upstream {
+
+SubsetSelector::SubsetSelector(const Protobuf::RepeatedPtrField<std::string>& selector_keys,
+                               envoy::config::cluster::v3::Cluster::LbSubsetConfig::
+                                   LbSubsetSelector::LbSubsetSelectorFallbackPolicy fallback_policy,
+                               const Protobuf::RepeatedPtrField<std::string>& fallback_keys_subset,
+                               bool single_host_per_subset)
+    : selector_keys_(selector_keys.begin(), selector_keys.end()),
+      fallback_keys_subset_(fallback_keys_subset.begin(), fallback_keys_subset.end()),
+      fallback_policy_(fallback_policy), single_host_per_subset_(single_host_per_subset) {
+
+  if (fallback_policy_ !=
+      envoy::config::cluster::v3::Cluster::LbSubsetConfig::LbSubsetSelector::KEYS_SUBSET) {
+    // defining fallback_keys_subset_ for a fallback policy other than KEYS_SUBSET doesn't have
+    // any effect and it is probably a user mistake. We should let the user know about it.
+    if (!fallback_keys_subset_.empty()) {
+      throwEnvoyExceptionOrPanic(
+          "fallback_keys_subset can be set only for KEYS_SUBSET fallback_policy");
+    }
+    return;
+  }
+
+  // if KEYS_SUBSET fallback policy is selected, fallback_keys_subset must not be empty, because
+  // it would be the same as not defining fallback policy at all (global fallback policy would be
+  // used)
+  if (fallback_keys_subset_.empty()) {
+    throwEnvoyExceptionOrPanic("fallback_keys_subset cannot be empty");
```

Comment:
so part of my goal of moving these out was being able to remove the exception code from E-M

could you change these to throw EnvoyException("...")
and add subset_lb.h to mobile/test/performance/files_em_does_not_use ?